syntax/text/parser/src/Luna/Syntax/Text/Parser/Parsing.hs

```diff
@@ -919,20 +921,98 @@ impSrc = buildAsg $ (\s -> liftIRBApp0 $ IR.unresolvedImpSrc' s) <$> (wrd <|> re
     rel = Import.Relative <$  symbol Lexer.Accessor <*> qualConsName
     abs = Import.Absolute <$> qualConsName
 
+------------------------------------
+-- === Foreign Import Parsing === --
+------------------------------------
+
+foreignImportList :: AsgParser SomeExpr
+foreignImportList = buildAsg $
+    (\lang imports ->
+        liftAstApp1 (IR.foreignImpList' lang) (sequence imports))
+    <$  (symbol Lexer.KwForeign *> symbol Lexer.KwImport)
+    <*> foreignLangName
+    <*  symbol Lexer.BlockStart
+    <*> discover (nonEmptyBlock' foreignLocationImportList)
+
+foreignLocationImportList :: AsgParser SomeExpr
+foreignLocationImportList = buildAsg $
+    (\loc imports ->
+        liftAstApp2 IR.foreignLocationImpList' loc (sequence imports))
+    <$> stringOrVarName
+    <*  symbol Lexer.BlockStart
+    <*> discover (nonEmptyBlock' foreignSymbolImport)
+
+foreignSymbolImport :: AsgParser SomeExpr
+foreignSymbolImport = buildAsg $ withRecovery recover
+    $   try (foreignSymbolImportWithSafety defaultFISafety)
+    <|> foreignSymbolImportWithSafety specifiedFISafety
+    where
+        recover = (\e ->
+            invalid "Invalid safety specification."
+            <$ Loc.unregisteredDropSymbolsUntil
+            (`elem` [Lexer.ETX, Lexer.EOL]))
+
+foreignSymbolImportWithSafety
+    :: AsgParser SomeExpr
+    -> IRParser SomeExpr
+foreignSymbolImportWithSafety getSafety =
+    (\safety foreignName localName importType ->
+        liftAstApp3
+        (foreignSymbolProxy localName)
+        safety
+        foreignName
+        importType)
+    <$> getSafety
+    <*> stringOrVarName
+    <*> funcName
+    <*  symbol Lexer.Typed
+    <*> valExpr
+    where
+        foreignSymbolProxy a b c d = IR.foreignSymbolImp' b c a d
+
```

Comment:
I would really try to keep the code vertically much more minimal. The `liftAstApp3` is unfortunate, because it doesnt allow for currying, so for now we have to use it like in your code, however all applicatives could be "one-lined". Otherwise its just like writing each function argument in a separate line. I'd prefer to shorten arg names then just use 12 lines instead of 3. I understand why reading over 80 chars is hard, but I strongly feel the same applies to reading just long vertical code.

```
foreignSymbolImportWithSafety getSafety = 
    (\safety foreignName locName impTp -> 
        liftAstApp3 (foreignSymbolProxy locName) safety foreignName impTp)
    <$> getSafety <*> stringOrVarName <*> func <* symbol Lexer.Typed <*> valExpr
```